File: modules/download_live.py
import os
import subprocess
import concurrent.futures
import time
import datetime
import re
import logging
import shutil
from modules.SharedDataStore import shared_stream_links_store


logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
BINARIES_DIR = os.path.join(BASE_DIR, 'binaries')
FFMPEG_BIN_PATH = os.path.join(BINARIES_DIR, 'ffmpeg.exe')
LOCK_FILES_DIR = os.path.join(BASE_DIR, 'lock_files')
SEGMENTS_DIR = os.path.join(BASE_DIR, 'segments')
VIDEOS_DIR = os.path.join(BASE_DIR, 'videos')

for lock_file in os.listdir(LOCK_FILES_DIR):
    os.remove(os.path.join(LOCK_FILES_DIR, lock_file))

def get_stream_links():
    return [(username, shared_stream_links_store.get_value(username)) for username in shared_stream_links_store.data]


def extract_stream_id(url):
    match = re.search(r'stream-(\d+)_', url)
    return match.group(1) if match else 'unknownid'

def concatenate_segments(username, stream_id):
    user_segment_dir = os.path.join(SEGMENTS_DIR, username)
    segments = [f for f in os.listdir(user_segment_dir) if f.startswith(f"{username}_{stream_id}") and f.endswith(".mp4")]
    
    if len(segments) == 1:
        current_date = datetime.datetime.now().strftime("%Y-%m-%d")
        single_segment_path = os.path.join(user_segment_dir, segments[0])
        output_file = os.path.join(VIDEOS_DIR, f"{username}_{stream_id}_{current_date}.mp4")
        shutil.copy(single_segment_path, output_file)
        logging.info(f"Copied single segment for {username} with stream ID {stream_id}")
    elif len(segments) > 1:
        list_file_path = os.path.join(user_segment_dir, f"{stream_id}_list.txt")
        with open(list_file_path, 'w') as list_file:
            for segment_file in sorted(segments):
                list_file.write(f"file '{os.path.join(user_segment_dir, segment_file)}'\n")

        current_date = datetime.datetime.now().strftime("%Y-%m-%d")
        output_file = os.path.join(VIDEOS_DIR, f"{username}_{stream_id}_{current_date}.mp4")
        ffmpeg_cmd = [
            FFMPEG_BIN_PATH, '-f', 'concat', '-safe', '0',
            '-i', list_file_path, '-c', 'copy', '-y', output_file
        ]
        subprocess.run(ffmpeg_cmd, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        logging.info(f"Concatenation completed for {username} with stream ID {stream_id}")
    else:
        logging.info(f"No segments found for {username} with stream ID {stream_id}, nothing to concatenate or copy.")


def download_livestream(username, stream_link):
    lock_file_path = os.path.join(LOCK_FILES_DIR, f'{username}.lock')
    user_segment_dir = os.path.join(SEGMENTS_DIR, username)
    os.makedirs(user_segment_dir, exist_ok=True)
    stream_id = extract_stream_id(stream_link)
    current_time = datetime.datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
    segment_path = os.path.join(user_segment_dir, f"{username}_{stream_id}_{current_time}.mp4")

    ffmpeg_cmd = [
        FFMPEG_BIN_PATH, '-i', stream_link,
        '-reconnect', '1', '-reconnect_at_eof', '1',
        '-reconnect_streamed', '1', '-reconnect_delay_max', '1', '-timeout', '30000000',
        '-c', 'copy', '-bsf:a', 'aac_adtstoasc', '-y', segment_path
    ]

    try:
        with open(lock_file_path, 'w') as lock_file:
            lock_file.write('')
        subprocess.run(ffmpeg_cmd, check=True, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        logging.info(f"Download completed for {username}.")
    except subprocess.CalledProcessError as e:
        logging.error(f"FFmpeg process error for {username}: {e}")
    finally:
        if os.path.exists(lock_file_path):
            concatenate_segments(username, stream_id)
            os.remove(lock_file_path)
            logging.info(f"Lock file removed and concatenation attempted for {username}.")


def main():
    with concurrent.futures.ThreadPoolExecutor() as executor:
        while True:
            logging.info("Checking for new stream links")
            stream_links = get_stream_links()
            logging.info(f"Found {len(stream_links)} stream links.")

            for username, stream_link in stream_links:
                if stream_link and not os.path.exists(os.path.join(LOCK_FILES_DIR, f'{username}.lock')):
                    executor.submit(download_livestream, username, stream_link)

            time.sleep(3)

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        logging.info("Script execution stopped by user.")
    except Exception as e:
        logging.critical(f"Critical error, stopping script: {e}")

File: src/bin/download_live.rs
use chrono::Local;
use regex::Regex;
use std::collections::HashMap;
use std::env;
use std::fs::{self, File};
use std::io::{self, Read, Write};
use std::path::{Path, PathBuf};
use std::process::{Command, Stdio};
use std::sync::Arc;
use tokio::sync::Mutex;
use tokio::time::{sleep, Duration};
use tokio::runtime::Builder;

static STREAM_LINKS_JSON_PATH: &str = "../json/stream_links.json";
static LOCK_FILES_DIR: &str = "../lock_files";
static SEGMENTS_DIR: &str = "../segments";
static VIDEOS_DIR: &str = "../videos";

async fn clear_lock_files(directory: &str) -> std::io::Result<()> {
    let paths = fs::read_dir(directory)?;
    for path in paths {
        let path = path?.path();
        if path.is_file() && path.extension().and_then(|s| s.to_str()) == Some("lock") {
            fs::remove_file(path)?;
        }
    }
    Ok(())
}

async fn read_stream_links() -> Result<HashMap<String, String>, Box<dyn std::error::Error>> {
    let mut contents = String::new();
    File::open(STREAM_LINKS_JSON_PATH)?.read_to_string(&mut contents)?;
    Ok(serde_json::from_str(&contents)?)
}

async fn download_livestream(username: &str, stream_link: &str, lock: Arc<Mutex<()>>, ffmpeg_path: &Path) -> Result<(), Box<dyn std::error::Error>> {
    let _lock = lock.lock().await;
    let lock_file_path = format!("{}/{}.lock", LOCK_FILES_DIR, username);
    let user_segment_dir = format!("{}/{}", SEGMENTS_DIR, username);
    fs::create_dir_all(&user_segment_dir)?;

    let stream_id = extract_stream_id(stream_link);
    let datetime = Local::now().format("%Y-%m-%d_%H-%M-%S").to_string();
    let segment_path = format!("{}/{}_{}_{}.mp4", user_segment_dir, username, stream_id, datetime);

    let status = Command::new(ffmpeg_path)
        .arg("-i")
        .arg(stream_link)
        .arg("-c")
        .arg("copy")
        .arg("-bsf:a")
        .arg("aac_adtstoasc")
        .arg("-y")
        .arg(&segment_path)
        .stdout(Stdio::null())
        .stderr(Stdio::null())
        .status()?;
    
    if status.success() {
        sleep(Duration::from_secs(1)).await;
        concatenate_segments(&user_segment_dir, username, &stream_id, ffmpeg_path)?;
        sleep(Duration::from_secs(1)).await;
    }

    fs::remove_file(lock_file_path)?;
    Ok(())
}

fn extract_stream_id(url: &str) -> String {
    Regex::new(r"stream-(\d+)_")
        .unwrap()
        .captures(url)
        .and_then(|caps| caps.get(1).map(|m| m.as_str().to_string()))
        .unwrap_or_else(|| "unknownid".to_string())
}

fn concatenate_segments(user_segment_dir: &str, username: &str, stream_id: &str, ffmpeg_path: &Path) -> io::Result<()> {
    let paths = fs::read_dir(user_segment_dir)?
        .filter_map(|entry| entry.ok())
        .map(|entry| entry.path())
        .filter(|path| path.is_file() && path.display().to_string().contains(stream_id) && path.extension().unwrap_or_default() == "mp4")
        .collect::<Vec<_>>();

    if paths.len() > 0 {
        let concat_file_path = format!("{}/{}_{}_concat.txt", user_segment_dir, username, stream_id);
        let mut concat_file = File::create(&concat_file_path)?;

        let _now = Local::now().format("%Y-%m-%d %H:%M:%S").to_string();

        for path in &paths {
            writeln!(concat_file, "file '{}'", path.canonicalize()?.display().to_string().replace('\\', "/"))?;
        }

        if paths.len() > 1 {
            let datetime = Local::now().format("%Y-%m-%d").to_string();
            let output_path = format!("{}/{}_{}_{}.mp4", VIDEOS_DIR, username, stream_id, datetime);
            let status = Command::new(ffmpeg_path)
                .arg("-f")
                .arg("concat")
                .arg("-safe")
                .arg("0")
                .arg("-i")
                .arg(&concat_file_path)
                .arg("-c")
                .arg("copy")
                .arg("-y")
                .arg(&output_path)
                .stdout(Stdio::null())
                .stderr(Stdio::null())
                .status()?;

            if !status.success() {
                let now = Local::now().format("%Y-%m-%d %H:%M:%S").to_string();
                println!("{} - ERROR - Error concatenating videos for user {} with stream id {}", now, username, stream_id);
            }
        } else {
            let datetime = Local::now().format("%Y-%m-%d").to_string();
            let single_output_path = format!("{}/{}_{}_{}.mp4", VIDEOS_DIR, username, stream_id, datetime);
            fs::copy(&paths[0], &single_output_path)?;
            let _now = Local::now().format("%Y-%m-%d %H:%M:%S").to_string();
        }
    } else {
        let now = Local::now().format("%Y-%m-%d %H:%M:%S").to_string();
        println!("{} - ERROR - No valid segments found for user {} with stream id {}", now, username, stream_id);
    }

    Ok(())
}

fn current_exe_path() -> Result<PathBuf, Box<dyn std::error::Error>> {
    Ok(env::current_exe()?)
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let lock_files_dir = "../lock_files";
    let segments_dir = "../segments";
    let videos_dir = "../videos";

    // Clear .lock files from lock_files directory
    if let Err(e) = clear_lock_files(lock_files_dir).await {
        eprintln!(" - ERROR - Failed to clear lock files: {}", e);
    }

    fs::create_dir_all(lock_files_dir)?;
    fs::create_dir_all(segments_dir)?;
    fs::create_dir_all(videos_dir)?;

    let runtime = Builder::new_multi_thread()
        .worker_threads(128)
        .enable_all()
        .build()
        .unwrap();

    loop {
        match read_stream_links().await {
            Ok(links) => {
                for (username, stream_link) in links.iter() {
                    let lock_file_path = format!("{}/{}.lock", lock_files_dir, username);
                    if !Path::new(&lock_file_path).exists() {
                        let now = Local::now().format("%Y-%m-%d %H:%M:%S").to_string();
                        println!("{} - INFO - Downloading livestream for user: {}", now, username);
                        if let Err(e) = fs::write(&lock_file_path, "") {
                            eprintln!("- ERROR - Error creating lock file for user {}: {}", username, e);
                            continue;
                        }
                        let lock = Arc::new(Mutex::new(()));
                        let username = username.clone();
                        let stream_link = stream_link.clone();
                        let handle = runtime.handle().clone();
                        let ffmpeg_path = current_exe_path()?.parent().unwrap().join("ffmpeg.exe");
                        handle.spawn(async move {
                            if let Err(e) = download_livestream(&username, &stream_link, lock, &ffmpeg_path).await {
                                eprintln!(" - ERROR - failed to download livestream for user {}: {}", username, e);
                            } else {
                                let now = Local::now().format("%Y-%m-%d %H:%M:%S").to_string();
                                println!("{} - INFO - Livestream downloaded successfully for user: {}", now, username);
                            }
                        });
                        sleep(Duration::from_secs(1)).await;
                    }
                }
            }
            Err(e) => {
                eprintln!(" - ERROR - Failed to read stream links: {}", e);
                sleep(Duration::from_secs(3)).await;
                continue; // Skip this loop iteration and try again
            }
        }
        sleep(Duration::from_secs(3)).await;
    }
}
